### PR TITLE
feat: added query validator `validateNestedQueriesHaveAValidRefOp()`

### PR DESCRIPTION
**Pull request by jimezesinachi**
# SynthQL pull request template

## Why

Added a validator function `validateNestedQueriesHaveAValidRefOp()` that checks the shape of nested queries to make sure they have a `RefOp` where clause (i.e JOIN clause) included

## What changed

`queries` and some tests in `backend`

## Important

Also includes the small change for the CLI schema validation

### CODE REVIEW

File: packages/queries/src/QueryBuilderError.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            'You need to add one like:',
+            '',
+            `.where({ columnName: col('rootQueryTableName.columnName') })`,
+            '',
+            'to your SynthQL query builder,',
+            'to connect (i.e. `JOIN`) the nested query to its parent query',
+            '',
+        ];
+
+        return new QueryBuilderError(type, lines.join('\n'));
```
[the remark this answers is not in the document]
Comment (fhur) [reply packages/queries/src/QueryBuilderError.ts:47]: Or maybe even simpler:

The table "parent" is including table "child", but "child" is missing a join predicate. Hint: are you missing `.where({some_id: "parent.some_id"})`

Where parent and child are replaces by actual table names

File: packages/queries/src/query.test.ts [hunk 2]
```diff
@@ -92,24 +92,6 @@ describe('queries', () => {
         };
     });
 
-    test('Find one actor by ID', () => {
```
Comment (fhur) [inline packages/queries/src/query.test.ts:95]: why did you remove tjos?

File: packages/queries/src/validators/validateNestedQueriesHaveAValidRefOp.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import { QueryBuilderError } from '../QueryBuilderError';
+import { Query } from '../types/types';
+import { isRefOp } from '../types/isRefOp';
+
+/**
```
Comment (fhur) [inline packages/queries/src/validators/validateNestedQueriesHaveAValidRefOp.ts:5]: Lots of grammar errors here.

```suggestion
Validate that every included sub-query has at least one refOp
```

File: packages/queries/src/validators/validateNestedQueriesHaveAValidRefOp.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ nested queries (includes), also has the
+ correct RefOp where clauses for each
+ */
+export function validateNestedQueriesHaveAValidRefOp<DB>(query: Query<DB>) {
+    const nestedQueries = Object.entries(query.include ?? {});
+
+    for (const [_, nestedQuery] of nestedQueries) {
+        const whereClauses = Object.entries(nestedQuery.where);
+
+        // Checking for one condition here:
```
Comment (fhur) [inline packages/queries/src/validators/validateNestedQueriesHaveAValidRefOp.ts:16]: this comment seems outdated. Please remove.

File: packages/queries/src/QueryBuilderError.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (fhur) [reply packages/queries/src/QueryBuilderError.ts:47]: Concrete example:

The table movie is including table actor, but actor is missing a join predicate. Hint: are you missing `where({some_id: "movie.some_id"})` on the actor query?

File: packages/queries/src/query.test.ts (hunk 2 above)
Comment (jimezesinachi) [reply packages/queries/src/query.test.ts:95]: Two identical tests... compare it to the one just above it

Comment (fhur) [reply packages/queries/src/query.test.ts:95]: Cool

File: packages/queries/src/QueryBuilderError.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        // this now, is that `AnyQuery`, still lives in the backend package
+        // public query: AnyQuery,
+        public type: string,
+        public message: string,
+    ) {
+        super(message);
+        Error.captureStackTrace(this, QueryBuilderError);
+    }
+
+    static createNestedQueryMissingRefOpWhereClauseError<DB>({
```
Comment (fhur) [inline packages/queries/src/QueryBuilderError.ts:16]: wait why do we have this and `createNestedQueryInvalidRefOpWhereClauseError`?

File: packages/queries/src/validators/validateNestedQueriesHaveAValidRefOp.test.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import { describe, expect, test } from 'vitest';
+import { col } from '../col';
+import { from } from '../generated';
+
+describe('validateNestedQueriesHaveAValidRefO', () => {
```
Comment (fhur) [inline packages/queries/src/validators/validateNestedQueriesHaveAValidRefOp.test.ts:5]:
```suggestion
describe('validateNestedQueriesHaveAValidRefOp', () => {
```

File: packages/queries/src/query.test.ts [hunk 7]
```diff
@@ -148,6 +131,7 @@ describe('queries', () => {
                 first_name: string;
                 last_name: string;
             }>;
+            filmActor: {}[];
```
Comment (fhur) [inline packages/queries/src/query.test.ts:134]: this shouldn't be there? were you getting a type error?

Comment (fhur) [reply packages/queries/src/query.test.ts:134]: queries with an empty select should not show up in the query result

File: packages/queries/src/QueryBuilderError.ts (hunk 5 above)
Comment (jimezesinachi) [reply packages/queries/src/QueryBuilderError.ts:16]: Invalid was one I created when I was working with the idea that we could check for incorrect RefOps... might still be useful in the future, so I kept it

File: packages/queries/src/query.test.ts (hunk 7 above)
Comment (jimezesinachi) [reply packages/queries/src/query.test.ts:134]: Nope. I just saw it was missing and added it.

File: packages/queries/src/QueryBuilderError.ts (hunk 5 above)
Comment (fhur) [reply packages/queries/src/QueryBuilderError.ts:16]: Don't, we can always re-build it. Unused code has a tendency to just take up space and never get used.

File: packages/queries/src/query.test.ts (hunk 7 above)
Comment (jimezesinachi) [reply packages/queries/src/query.test.ts:134]: > queries with an empty select should not show up in the query result

Ahh, I see now. That suggests we have a bug then

File: packages/queries/src/validators/validateNestedQueriesHaveAValidRefOp.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import { QueryBuilderError } from '../QueryBuilderError';
+import { Query } from '../types/types';
+import { isRefOp } from '../types/isRefOp';
+
+/**
+ Validate that every included sub-query has at least one RefOp
+ */
+export function validateNestedQueriesHaveAValidRefOp<DB>(query: Query<DB>) {
+    const nestedQueries = Object.entries(query.include ?? {});
```
Comment (fhur) [inline packages/queries/src/validators/validateNestedQueriesHaveAValidRefOp.ts:9]: just realised: you're using `entries` but only actually using the values. Same below. Just use `Object.values`.

More attention to details please :)

**Review by fhur — APPROVED** [review]
